client/boot/geolocation.ts

```diff
@@ -0,0 +1,31 @@
+import { setGeoLocation } from '@automattic/number-formatters';
+import defaultCalypsoI18n from 'i18n-calypso';
+import { type UserData } from 'calypso/lib/user/user';
+
+async function fetchCountryCode(): Promise< string > {
+	try {
+		const response = await fetch( 'https://public-api.wordpress.com/geo/' );
+		const data = await response.json();
+		return data.country_short ?? '';
+	} catch {
+		return '';
+	}
+}
+
+/**
+ * Initializes country code for localization and formatting libraries from the current user. If the
+ * user is not logged in, it will fetch the country code from the geolocation API.
+ * @param currentUser The current user.
+ */
+export async function setupCountryCode( currentUser: UserData | null ) {
+	const countryCode = currentUser?.user_ip_country_code ?? ( await fetchCountryCode() );
+	if ( ! countryCode ) {
+		return;
+	}
+
+	setGeoLocation( countryCode );
```

Comment:
This part is not reactive right? Meaning, if `countryCode` ends up coming async, it won't trigger a rerender per se / on its own. We update the singleton but nothing further happens. It's how it is now too, and not a big deal either. 

Or am I missing something?

Reply:
> It's how it is now too, and not a big deal either.

Actually, not sure if that's true. Since now (currently) we update the i18n instance, it triggers other side effects, but nothing that's relevant to number-formatters. 

I'd rather we rebase over https://github.com/Automattic/wp-calypso/pull/102878, which removes `i18n.geoLocation` as no longer used. Just to make sure we test this isolated i.e. with the only effect on number-formatters.

Reply:
Thanks. Indeed only emits change for locale changing. In any case, `geoLocation` was also ported from `@automattic/format-currency`, which is how it works there too.

We'll probably need to explore introducing reactivity to number-formatters. I'm thinking encapsulating these with a hook like `const { formatCurrency, ... } = useNumberFormatters();`, then decide where the context will live. A brief exploration points me to a singleton store with `useSyncExternalStore`, which would make it the more "conventional" way outside of using context provider. But all that would be like a number-formatters v2.